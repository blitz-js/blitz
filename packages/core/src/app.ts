/*
 * IF YOU CHANGE THIS FILE
 *    You also need to update the rewrite map in
 *    packages/babel-preset/src/rewrite-imports.ts
 */
export {default as App} from "next/app"
