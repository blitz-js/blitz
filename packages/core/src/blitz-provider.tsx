import React, {FC, useRef} from "react"
import {QueryClient, QueryClientProvider} from "react-query"
import {Hydrate, HydrateOptions} from "react-query/hydration"
import {queryClient} from "./utils/react-query-utils"

export type BlitzProviderProps = {
  client?: QueryClient
  contextSharing?: boolean
  dehydratedState?: unknown
  hydrateOptions?: HydrateOptions
}

export const BlitzProvider: FC<BlitzProviderProps> = ({
  client,
  contextSharing = false,
  dehydratedState,
  hydrateOptions,
  children,
}) => {
  const queryClientRef = useRef<QueryClient>()
  if (!queryClientRef.current) {
    queryClientRef.current = queryClient
  }

  return (
    <QueryClientProvider client={client ?? queryClient} contextSharing={contextSharing}>
      <Hydrate state={dehydratedState} options={hydrateOptions}>
        {children}
      </Hydrate>
    </QueryClientProvider>
  )
}
