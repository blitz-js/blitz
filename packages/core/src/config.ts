/*
 * IF YOU CHANGE THIS FILE
 *    You also need to update the rewrite map in
 *    packages/babel-preset/src/rewrite-imports.ts
 */
export {default as getConfig, setConfig} from "next/config"
