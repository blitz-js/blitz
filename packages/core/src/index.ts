export * from './useQuery'
export * from './ssrQuery'
export * from './rpc'

export {
  GetStaticProps,
  GetStaticPaths,
  GetServerSideProps,
  NextPage as BlitzPage,
  NextApiRequest as BlitzApiRequest,
  NextApiResponse as BlitzApiResponse,
} from 'next'

export {default as Head} from 'next/head'

export {default as Link} from 'next/link'

export {Router, useRouter, withRouter} from 'next/router'

export {
  default as Document,
  Html,
  Head as DocumentHead,
  Main,
  NextScript,
  DocumentContext,
} from 'next/document'

export {default as dynamic} from 'next/dynamic'
