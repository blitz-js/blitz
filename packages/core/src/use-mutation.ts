import {
  MutateOptions,
  useMutation as useReactQueryMutation,
  UseMutationOptions,
  UseMutationResult,
} from "react-query"
import {sanitizeMutation} from "./utils/react-query-utils"

/*
 * We have to override react-query's MutationFunction and MutationResultPair
 * types so because we have throwOnError:true by default. And by the RQ types
 * have the mutate function result typed as TData|undefined which isn't typed
 * properly with throwOnError.
 *
 * So this fixes that.
 */
export declare type MutateFunction<
  TData,
  TError = unknown,
  TVariables = unknown,
  TContext = unknown
> = (
  variables?: TVariables,
  config?: MutateOptions<TData, TError, TVariables, TContext>,
) => Promise<TData>

export declare type MutationResultPair<TData, TError, TVariables, TContext> = [
  MutateFunction<TData, TError, TVariables, TContext>,
  Omit<UseMutationResult<TData, TError>, "mutate" | "mutateAsync">,
]

export declare type MutationFunction<TData, TVariables = unknown> = (
  variables: TVariables,
  ctx?: any,
) => Promise<TData>

export function useMutation<
  TData = unknown,
  TError = unknown,
  TVariables = void,
  TContext = unknown
>(
  mutationResolver: MutationFunction<TData, TVariables>,
  config?: UseMutationOptions<TData, TError, TVariables, TContext>,
): MutationResultPair<TData, TError, TVariables, TContext> {
  const enhancedResolverRpcClient = sanitizeMutation(mutationResolver)

  const {mutate, mutateAsync, ...rest} = useReactQueryMutation<TData, TError, TVariables, TContext>(
    (variables) => enhancedResolverRpcClient(variables, {fromQueryHook: true}),
    {
      throwOnError: true,
      ...config,
    } as any,
  )

  return [mutateAsync, rest] as MutationResultPair<TData, TError, TVariables, TContext>
}
