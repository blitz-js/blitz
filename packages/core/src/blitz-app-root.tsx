import React, {ComponentPropsWithoutRef, useEffect} from "react"
import {useAuthorizeIf} from "./auth/auth-client"
import {publicDataStore} from "./auth/public-data-store"
import {BlitzProvider} from "./blitz-provider"
import {Head} from "./head"
import {AppProps, BlitzPage} from "./types"

const customCSS = `
  body::before {
    content: "";
    display: block;
    position: fixed;
    width: 100%;
    height: 100%;
    top: 0;
    left: 0;
    z-index: 99999;
    background-color: white;
  }

  .blitz-first-render-complete body::before {
    display: none;
  }
`
const noscriptCSS = `
  body::before {
    content: none
  }
`

const NoPageFlicker = () => {
  return (
    <Head>
      <style dangerouslySetInnerHTML={{__html: customCSS}} />
      <noscript>
        <style dangerouslySetInnerHTML={{__html: noscriptCSS}} />
      </noscript>
    </Head>
  )
}

export function withBlitzInnerWrapper(Page: BlitzPage) {
  const BlitzInnerRoot = (props: ComponentPropsWithoutRef<BlitzPage>) => {
    useAuthorizeIf(Page.authenticate === true)

    return <Page {...props} />
  }
  for (let [key, value] of Object.entries(Page)) {
    ;(BlitzInnerRoot as any)[key] = value
  }
  if (process.env.NODE_ENV !== "production") {
    BlitzInnerRoot.displayName = `BlitzInnerRoot`
  }
  return BlitzInnerRoot
}

export function withBlitzAppRoot(UserAppRoot: React.ComponentType<any>) {
  const BlitzOuterRoot = (props: AppProps) => {
    if (typeof window !== "undefined") {
      if (publicDataStore.getData().userId) {
        if (props.Component.redirectAuthenticatedTo) {
          window.location.replace(props.Component.redirectAuthenticatedTo)
        }
      } else {
        const authenticate = props.Component.authenticate
        if (
          authenticate &&
          typeof authenticate === "object" &&
          typeof authenticate.redirectTo === "string"
        ) {
          const url = new URL(authenticate.redirectTo, window.location.href)
          url.searchParams.append("next", window.location.pathname)
          window.location.replace(url.toString())
        }
      }
    }
    const component = React.useMemo(() => withBlitzInnerWrapper(props.Component), [props.Component])

    const noPageFlicker =
      props.Component.suppressFirstRenderFlicker ||
      props.Component.authenticate !== undefined ||
      props.Component.redirectAuthenticatedTo

    useEffect(() => {
      document.documentElement.classList.add("blitz-first-render-complete")
    }, [])

    return (
      <BlitzProvider dehydratedState={props.pageProps.dehydratedState}>
        {noPageFlicker && <NoPageFlicker />}
        <UserAppRoot {...props} Component={component} />
      </BlitzProvider>
    )
  }
  return BlitzOuterRoot
}
