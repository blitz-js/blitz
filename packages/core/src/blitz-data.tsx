import {getConfig} from "@blitzjs/config"
import htmlescape from "htmlescape"
import React from "react"
import {isClient} from "./utils"

export type BlitzRuntimeData = {
  suspenseEnabled: boolean
  sessionCookiePrefix: string
}

export function _getBlitzRuntimeData(): BlitzRuntimeData {
  const config = getConfig()
  return {
    sessionCookiePrefix: (config._meta.packageName || "blitz").replace(/[^a-zA-Z0-9-_]/g, "_"),
    suspenseEnabled: config.experimental?.reactRoot !== false,
  }
}

export function getBlitzRuntimeData() {
  if (isClient && !process.env.JEST_WORKER_ID) {
    return window.__BLITZ_DATA__
  } else {
    if (!global.__BLITZ_DATA__) {
      global.__BLITZ_DATA__ = _getBlitzRuntimeData()
    }
    return global.__BLITZ_DATA__
  }
}

// Automatically deserialize __BLITZ_DATA__ in a browser environment
if (isClient && !process.env.JEST_WORKER_ID) {
  deserializeAndSetBlitzDataOnWindow()
}

export function deserializeAndSetBlitzDataOnWindow() {
  try {
    const data: typeof window["__BLITZ_DATA__"] = JSON.parse(
      document.getElementById("__BLITZ_DATA__")!.textContent!,
    )
    window.__BLITZ_DATA__ = data
  } catch (e) {
    console.error(
      "Error deserializing __BLITZ__DATA__. Make sure you have a custom _document.js/tsx page that uses <BlitzScript/>",
      e,
    )
  }
}

export function BlitzData() {
  return (
    <script
      id="__BLITZ_DATA__"
      type="application/json"
      dangerouslySetInnerHTML={{
        __html: htmlescape(_getBlitzRuntimeData()),
      }}
    />
  )
}
