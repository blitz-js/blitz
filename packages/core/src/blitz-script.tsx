import {NextScript} from "next/document"
import React, {ComponentPropsWithoutRef} from "react"
import {BlitzData} from "./blitz-data"

export type BlitzScriptProps = ComponentPropsWithoutRef<typeof NextScript>

export function BlitzScript(props: BlitzScriptProps) {
  return (
    <>
      <BlitzData />
      <NextScript {...props} />
    </>
  )
}
