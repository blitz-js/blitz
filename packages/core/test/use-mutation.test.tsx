import React from "react"
import {useMutation} from "../src/use-mutation"
import {act, enhanceMutationFn, enhanceQueryFn, render, screen} from "./test-utils"

describe("useMutation", () => {
  const setupHook = (resolver: (...args: any) => Promise<any>): [{mutate?: Function}, Function] => {
    let res = {}
    function TestHarness() {
      const [mutate, {isSuccess}] = useMutation(resolver)
      Object.assign(res, {mutate})
      return <div id="harness">{isSuccess ? "Sent" : "Waiting"}</div>
    }

    const ui = () => <TestHarness />

    const {rerender} = render(ui())
    return [res, () => rerender(ui())]
  }

  describe("useMutation calls the resolver with the argument", () => {
    // eslint-disable-next-line require-await
    const mutateFn = jest.fn()
    it("should work with Blitz mutations", async () => {
      const [res] = setupHook(enhanceMutationFn(mutateFn))
      await act(async () => {
        await res.mutate!("data")
        await screen.findByText("Sent")
        expect(mutateFn).toHaveBeenCalledTimes(1)
        expect(mutateFn).toHaveBeenCalledWith("data", {fromQueryHook: true})
      })
    })

    it("shouldn't work with regular functions", () => {
      console.error = jest.fn()
      expect(() => setupHook(mutateFn)).toThrowErrorMatchingSnapshot()
    })

    it("shouldn't work with query function", () => {
      console.error = jest.fn()
      const mutationFn = jest.fn()

      expect(() => setupHook(enhanceQueryFn(mutationFn))).toThrowErrorMatchingSnapshot()
    })
  })
})
