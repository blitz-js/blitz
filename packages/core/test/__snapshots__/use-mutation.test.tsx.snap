// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useMutation useMutation calls the resolver with the argument shouldn't work with query function 1`] = `"\\"useMutation\\" was expected to be called with a mutation but was called with a \\"query\\""`;

exports[`useMutation useMutation calls the resolver with the argument shouldn't work with regular functions 1`] = `"Either the file path to your resolver is incorrect (must be in a \\"queries\\" or \\"mutations\\" folder that isn't nested inside \\"pages\\" or \\"api\\") or you are trying to use Blitz's useQuery to fetch from third-party APIs (to do that, import useQuery directly from \\"react-query\\")"`;
